File: src/services/spend.service.ts
import { Prisma, type ApprovalMode, type Pool, type Vote, type SpendRequest } from "@prisma/client";
import { prisma } from "../lib/prisma.js";
import { Errors } from "../utils/errors.js";
import { ensureActivePool } from "./pool.service.js";

type Resolution = "PENDING" | "APPROVED" | "REJECTED";

/**
 * Returns the new SpendRequest.status based on current votes & pool rules.
 * Pure function (no I/O) so it can be unit tested.
 */
export function resolveVotingStatus(
  pool: Pick<Pool, "approvalMode" | "approvalThreshold" | "createdById">,
  totalEligibleMembers: number,
  votes: Pick<Vote, "voterId" | "decision">[]
): Resolution {
  const approvals = votes.filter((v) => v.decision === "APPROVE").length;
  const rejections = votes.filter((v) => v.decision === "REJECT").length;
  const abstentions = votes.filter((v) => v.decision === "ABSTAIN").length;

  if (pool.approvalMode === ("ADMIN_ONLY" as ApprovalMode)) {
    if (votes.some((v) => v.voterId === pool.createdById && v.decision === "APPROVE")) return "APPROVED";
    if (votes.some((v) => v.voterId === pool.createdById && v.decision === "REJECT")) return "REJECTED";
    return "PENDING";
  }

  const denominator = Math.max(1, totalEligibleMembers - abstentions);

  let threshold: number;
  if (pool.approvalMode === ("UNANIMOUS" as ApprovalMode)) threshold = 100;
  else if (pool.approvalMode === ("MAJORITY" as ApprovalMode)) threshold = 51;
  else threshold = pool.approvalThreshold;

  const approveRatio = (approvals / denominator) * 100;
  if (approveRatio >= threshold) return "APPROVED";

  // Can the remaining unvoted members still push it over the threshold?
  const remaining = totalEligibleMembers - approvals - rejections - abstentions;
  const maxPossibleApprovals = approvals + remaining;
  const maxRatio = (maxPossibleApprovals / denominator) * 100;
  if (maxRatio < threshold) return "REJECTED";

  return "PENDING";
}

export async function createSpendRequest(input: {
  poolId: string;
  requesterId: string;
  amount: string;
  title: string;
  description?: string;
  category: SpendRequest["category"];
  receiptUrl?: string;
  isEmergency?: boolean;
  expiresInHours: number;
}) {
  return prisma.$transaction(async (tx) => {
    const pool = await tx.pool.findUnique({ where: { id: input.poolId } });
    if (!pool) throw Errors.notFound("Pool");
    ensureActivePool(pool);

    const member = await tx.poolMember.findUnique({
      where: { poolId_userId: { poolId: pool.id, userId: input.requesterId } },
    });
    if (!member || !member.isActive) throw Errors.forbidden("Not a member");

    const amount = new Prisma.Decimal(input.amount);
    if (pool.spendLimit && amount.greaterThan(pool.spendLimit)) {
      throw Errors.conflict(`Amount exceeds pool spend limit of RM${pool.spendLimit.toFixed(2)}`);
    }
    if (pool.currentBalance.lessThan(amount)) {
      throw Errors.conflict("Pool balance insufficient");
    }

    if (input.isEmergency && pool.type !== "FAMILY") {
      throw Errors.conflict("Emergency override only available for FAMILY pools");
    }
    if (input.isEmergency && !pool.emergencyOverride) {
      throw Errors.conflict("Emergency override is disabled for this pool");
    }

    const expiresAt = new Date(Date.now() + input.expiresInHours * 60 * 60 * 1000);

    const created = await tx.spendRequest.create({
      data: {
        poolId: pool.id,
        requesterId: input.requesterId,
        amount,
        title: input.title,
        description: input.description,
        category: input.category,
        receiptUrl: input.receiptUrl,
        isEmergency: input.isEmergency ?? false,
        status: input.isEmergency ? "APPROVED" : "PENDING",
        expiresAt,
        resolvedAt: input.isEmergency ? new Date() : null,
      },
    });

    const others = await tx.poolMember.findMany({
      where: { poolId: pool.id, isActive: true, NOT: { userId: input.requesterId } },
      select: { userId: true },
    });
    if (others.length > 0) {
      await tx.notification.createMany({
        data: others.map((m) => ({
          userId: m.userId,
          type: input.isEmergency ? "SPEND_REQUEST_APPROVED" : "SPEND_REQUEST_NEW",
          title: input.isEmergency ? "Emergency spend approved" : "Vote needed",
          body: `${input.title} — RM${amount.toFixed(2)}`,
          metadata: { poolId: pool.id, spendRequestId: created.id },
        })),
      });
    }

    return created;
  });
}

export async function castVote(input: {
  poolId: string;
  spendRequestId: string;
  voterId: string;
  decision: "APPROVE" | "REJECT" | "ABSTAIN";
  comment?: string;
}) {
  return prisma.$transaction(async (tx) => {
    const sr = await tx.spendRequest.findUnique({ where: { id: input.spendRequestId } });
    if (!sr || sr.poolId !== input.poolId) throw Errors.notFound("Spend request");
    if (sr.status !== "PENDING") throw Errors.voteClosed();
    if (sr.expiresAt < new Date()) {
      await tx.spendRequest.update({
        where: { id: sr.id },
        data: { status: "EXPIRED", resolvedAt: new Date() },
      });
      throw Errors.voteClosed();
    }
    if (sr.requesterId === input.voterId) throw Errors.forbidden("Requester cannot vote on own request");

    const member = await tx.poolMember.findUnique({
      where: { poolId_userId: { poolId: sr.poolId, userId: input.voterId } },
    });
    if (!member || !member.isActive) throw Errors.forbidden("Not a member");
    if (member.role === "VIEWER") throw Errors.forbidden("Viewers cannot vote");

    const existing = await tx.vote.findUnique({
      where: { spendRequestId_voterId: { spendRequestId: sr.id, voterId: input.voterId } },
    });
    if (existing) throw Errors.alreadyVoted();

    const vote = await tx.vote.create({
      data: {
        spendRequestId: sr.id,
        voterId: input.voterId,
        decision: input.decision,
        comment: input.comment,
      },
    });

    const pool = await tx.pool.findUniqueOrThrow({ where: { id: sr.poolId } });
    const eligibleMembers = await tx.poolMember.count({
      where: {
        poolId: sr.poolId,
        isActive: true,
        role: { not: "VIEWER" },
        NOT: { userId: sr.requesterId },
      },
    });
    const allVotes = await tx.vote.findMany({
      where: { spendRequestId: sr.id },
      select: { voterId: true, decision: true },
    });

    const resolution = resolveVotingStatus(pool, eligibleMembers, allVotes);

    let updated = sr;
    if (resolution !== "PENDING") {
      updated = await tx.spendRequest.update({
        where: { id: sr.id },
        data: { status: resolution, resolvedAt: new Date() },
      });
      const noteType = resolution === "APPROVED" ? "SPEND_REQUEST_APPROVED" : "SPEND_REQUEST_REJECTED";
      await tx.notification.create({
        data: {
          userId: sr.requesterId,
          type: noteType,
          title: `Spend request ${resolution.toLowerCase()}`,
          body: `${sr.title} — RM${sr.amount.toFixed(2)}`,
          metadata: { poolId: sr.poolId, spendRequestId: sr.id },
        },
      });
    }

    return { vote, spendRequest: updated, resolution };
  });
}

export async function executeApprovedSpend(input: {
  poolId: string;
  spendRequestId: string;
  actorId: string;
}) {
  return prisma.$transaction(async (tx) => {
    const sr = await tx.spendRequest.findUnique({ where: { id: input.spendRequestId } });
    if (!sr || sr.poolId !== input.poolId) throw Errors.notFound("Spend request");
    if (sr.status !== "APPROVED") throw Errors.conflict(`Cannot execute spend in ${sr.status} state`);
    if (sr.requesterId !== input.actorId) {
      const actor = await tx.poolMember.findUnique({
        where: { poolId_userId: { poolId: sr.poolId, userId: input.actorId } },
      });
      if (!actor || (actor.role !== "OWNER" && actor.role !== "ADMIN")) {
        throw Errors.forbidden("Only requester or pool admin can execute");
      }
    }

    const pool = await tx.pool.findUniqueOrThrow({ where: { id: sr.poolId } });
    if (pool.isFrozen) throw Errors.poolFrozen();
    if (pool.currentBalance.lessThan(sr.amount)) throw Errors.conflict("Pool balance insufficient at execution time");

    const requester = await tx.user.findUniqueOrThrow({ where: { id: sr.requesterId } });

    const poolBefore = pool.currentBalance;
    const poolAfter = poolBefore.minus(sr.amount);
    const userBefore = requester.mainBalance;
    const userAfter = userBefore.plus(sr.amount);

    await tx.pool.update({ where: { id: pool.id }, data: { currentBalance: poolAfter } });
    await tx.user.update({ where: { id: requester.id }, data: { mainBalance: userAfter } });

    await tx.transaction.create({
      data: {
        poolId: pool.id,
        userId: requester.id,
        type: "SPEND",
        direction: "OUT",
        amount: sr.amount,
        balanceBefore: poolBefore,
        balanceAfter: poolAfter,
        description: `Spend: ${sr.title}`,
        spendRequestId: sr.id,
      },
    });
    await tx.transaction.create({
      data: {
        userId: requester.id,
        type: "SPEND",
        direction: "IN",
        amount: sr.amount,
        balanceBefore: userBefore,
        balanceAfter: userAfter,
        description: `Pool payout: ${sr.title}`,
        spendRequestId: sr.id,
        metadata: { poolId: pool.id },
      },
    });

    const updated = await tx.spendRequest.update({
      where: { id: sr.id },
      data: { status: "EXECUTED" },
    });

    return { spendRequest: updated, poolBalance: poolAfter, userBalance: userAfter };
  });
}

export async function cancelSpendRequest(input: { poolId: string; spendRequestId: string; actorId: string }) {
  const sr = await prisma.spendRequest.findUnique({ where: { id: input.spendRequestId } });
  if (!sr || sr.poolId !== input.poolId) throw Errors.notFound("Spend request");
  if (sr.requesterId !== input.actorId) throw Errors.forbidden("Only requester can cancel");
  if (sr.status !== "PENDING") throw Errors.conflict(`Cannot cancel spend in ${sr.status} state`);
  return prisma.spendRequest.update({
    where: { id: sr.id },
    data: { status: "CANCELLED", resolvedAt: new Date() },
  });
}

/** Sweep job — should be called periodically to expire stale pending requests. */
export async function expireStaleRequests() {
  const now = new Date();
  const stale = await prisma.spendRequest.findMany({
    where: { status: "PENDING", expiresAt: { lt: now } },
    select: { id: true, requesterId: true, poolId: true, title: true },
  });
  if (stale.length === 0) return 0;
  await prisma.$transaction([
    prisma.spendRequest.updateMany({
      where: { id: { in: stale.map((s) => s.id) } },
      data: { status: "EXPIRED", resolvedAt: now },
    }),
    prisma.notification.createMany({
      data: stale.map((s) => ({
        userId: s.requesterId,
        type: "SPEND_REQUEST_EXPIRED" as const,
        title: "Spend request expired",
        body: s.title,
        metadata: { poolId: s.poolId, spendRequestId: s.id },
      })),
    }),
  ]);
  return stale.length;
}

File: backend/app/services/spend_service.py
"""Spend-request flow — create / vote / cancel / execute / expire-sweep."""
from datetime import datetime, timedelta, timezone
from decimal import Decimal
from typing import Any, Iterable, Literal

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from ..enums import (
    ApprovalMode, MemberRole, NotificationType, PoolType, SpendStatus,
    TransactionDirection, TransactionType, VoteDecision,
)
from ..errors import Errors
from ..models import Notification, Pool, PoolMember, SpendRequest, Transaction, User, Vote
from .pool_service import ensure_active_pool

Resolution = Literal["PENDING", "APPROVED", "REJECTED"]


def resolve_voting_status(
    *, approval_mode: ApprovalMode, approval_threshold: int, created_by_id: str,
    total_eligible_members: int, votes: Iterable[dict],
) -> Resolution:
    """Pure function: derive resolution from current vote tally + pool rules.

    Mirrors src/services/spend.service.ts#resolveVotingStatus.
    """
    votes = list(votes)
    approvals = sum(1 for v in votes if v["decision"] == "APPROVE")
    rejections = sum(1 for v in votes if v["decision"] == "REJECT")
    abstentions = sum(1 for v in votes if v["decision"] == "ABSTAIN")

    if approval_mode == ApprovalMode.ADMIN_ONLY:
        for v in votes:
            if v["voterId"] == created_by_id:
                if v["decision"] == "APPROVE":
                    return "APPROVED"
                if v["decision"] == "REJECT":
                    return "REJECTED"
        return "PENDING"

    denominator = max(1, total_eligible_members - abstentions)
    if approval_mode == ApprovalMode.UNANIMOUS:
        threshold = 100
    elif approval_mode == ApprovalMode.MAJORITY:
        threshold = 51
    else:
        threshold = approval_threshold

    approve_ratio = (approvals / denominator) * 100
    if approve_ratio >= threshold:
        return "APPROVED"

    remaining = total_eligible_members - approvals - rejections - abstentions
    max_possible = approvals + remaining
    max_ratio = (max_possible / denominator) * 100
    if max_ratio < threshold:
        return "REJECTED"
    return "PENDING"


async def create_spend_request(session: AsyncSession, *, pool_id: str, requester_id: str,
                               amount: str, title: str, description: str | None,
                               category: str, receipt_url: str | None,
                               is_emergency: bool, expires_in_hours: int) -> SpendRequest:
    pool = (await session.execute(select(Pool).where(Pool.id == pool_id))).scalar_one_or_none()
    if not pool:
        raise Errors.not_found("Pool")
    ensure_active_pool(pool)
    member = (await session.execute(
        select(PoolMember).where(PoolMember.poolId == pool_id, PoolMember.userId == requester_id)
    )).scalar_one_or_none()
    if not member or not member.isActive:
        raise Errors.forbidden("Not a member")

    amt = Decimal(amount)
    if pool.spendLimit is not None and amt > pool.spendLimit:
        raise Errors.conflict(f"Amount exceeds pool spend limit of RM{pool.spendLimit:.2f}")
    if pool.currentBalance < amt:
        raise Errors.conflict("Pool balance insufficient")
    if is_emergency and pool.type != PoolType.FAMILY:
        raise Errors.conflict("Emergency override only available for FAMILY pools")
    if is_emergency and not pool.emergencyOverride:
        raise Errors.conflict("Emergency override is disabled for this pool")

    expires_at = datetime.now(timezone.utc) + timedelta(hours=expires_in_hours)
    sr = SpendRequest(
        poolId=pool.id, requesterId=requester_id, amount=amt, title=title,
        description=description, category=category, receiptUrl=receipt_url,
        isEmergency=is_emergency,
        status=SpendStatus.APPROVED if is_emergency else SpendStatus.PENDING,
        expiresAt=expires_at,
        resolvedAt=datetime.now(timezone.utc) if is_emergency else None,
    )
    session.add(sr)
    await session.flush()

    others = (await session.execute(
        select(PoolMember.userId).where(
            PoolMember.poolId == pool.id, PoolMember.isActive.is_(True),
            PoolMember.userId != requester_id,
        )
    )).scalars().all()
    nt = NotificationType.SPEND_REQUEST_APPROVED if is_emergency else NotificationType.SPEND_REQUEST_NEW
    note_title = "Emergency spend approved" if is_emergency else "Vote needed"
    for uid in others:
        session.add(Notification(
            userId=uid, type=nt, title=note_title,
            body=f"{title} — RM{amt:.2f}",
            metadata_={"poolId": pool.id, "spendRequestId": sr.id},
        ))
    await session.commit()
    await session.refresh(sr)
    return sr


async def cast_vote(session: AsyncSession, *, pool_id: str, spend_request_id: str,
                    voter_id: str, decision: str, comment: str | None) -> dict[str, Any]:
    sr = (await session.execute(
        select(SpendRequest).where(SpendRequest.id == spend_request_id)
    )).scalar_one_or_none()
    if not sr or sr.poolId != pool_id:
        raise Errors.not_found("Spend request")
    if sr.status != SpendStatus.PENDING:
        raise Errors.vote_closed()
    now = datetime.now(timezone.utc)
    if sr.expiresAt < now:
        sr.status = SpendStatus.EXPIRED
        sr.resolvedAt = now
        await session.commit()
        raise Errors.vote_closed()
    if sr.requesterId == voter_id:
        raise Errors.forbidden("Requester cannot vote on own request")

    member = (await session.execute(
        select(PoolMember).where(PoolMember.poolId == sr.poolId, PoolMember.userId == voter_id)
    )).scalar_one_or_none()
    if not member or not member.isActive:
        raise Errors.forbidden("Not a member")
    if member.role == MemberRole.VIEWER:
        raise Errors.forbidden("Viewers cannot vote")

    existing = (await session.execute(
        select(Vote).where(Vote.spendRequestId == sr.id, Vote.voterId == voter_id)
    )).scalar_one_or_none()
    if existing:
        raise Errors.already_voted()

    vote = Vote(spendRequestId=sr.id, voterId=voter_id, decision=decision, comment=comment)
    session.add(vote)
    await session.flush()

    pool = (await session.execute(select(Pool).where(Pool.id == sr.poolId))).scalar_one()
    eligible = (await session.execute(
        select(func.count()).select_from(PoolMember).where(
            PoolMember.poolId == sr.poolId,
            PoolMember.isActive.is_(True),
            PoolMember.role != MemberRole.VIEWER,
            PoolMember.userId != sr.requesterId,
        )
    )).scalar_one()
    all_votes = (await session.execute(
        select(Vote.voterId, Vote.decision).where(Vote.spendRequestId == sr.id)
    )).all()
    vote_dicts = [{"voterId": vid, "decision": d.value if hasattr(d, "value") else d} for vid, d in all_votes]

    resolution = resolve_voting_status(
        approval_mode=pool.approvalMode, approval_threshold=pool.approvalThreshold,
        created_by_id=pool.createdById, total_eligible_members=int(eligible), votes=vote_dicts,
    )

    if resolution != "PENDING":
        sr.status = SpendStatus.APPROVED if resolution == "APPROVED" else SpendStatus.REJECTED
        sr.resolvedAt = datetime.now(timezone.utc)
        nt = (
            NotificationType.SPEND_REQUEST_APPROVED if resolution == "APPROVED"
            else NotificationType.SPEND_REQUEST_REJECTED
        )
        session.add(Notification(
            userId=sr.requesterId, type=nt,
            title=f"Spend request {resolution.lower()}",
            body=f"{sr.title} — RM{sr.amount:.2f}",
            metadata_={"poolId": sr.poolId, "spendRequestId": sr.id},
        ))

    await session.commit()
    await session.refresh(sr)
    await session.refresh(vote)
    return {"vote": vote, "spendRequest": sr, "resolution": resolution}


async def execute_approved_spend(session: AsyncSession, *, pool_id: str,
                                 spend_request_id: str, actor_id: str) -> dict[str, Any]:
    sr = (await session.execute(
        select(SpendRequest).where(SpendRequest.id == spend_request_id)
    )).scalar_one_or_none()
    if not sr or sr.poolId != pool_id:
        raise Errors.not_found("Spend request")
    if sr.status != SpendStatus.APPROVED:
        raise Errors.conflict(f"Cannot execute spend in {sr.status.value} state")
    if sr.requesterId != actor_id:
        actor = (await session.execute(
            select(PoolMember).where(PoolMember.poolId == sr.poolId, PoolMember.userId == actor_id)
        )).scalar_one_or_none()
        if not actor or actor.role not in (MemberRole.OWNER, MemberRole.ADMIN):
            raise Errors.forbidden("Only requester or pool admin can execute")

    pool = (await session.execute(select(Pool).where(Pool.id == sr.poolId))).scalar_one()
    if pool.isFrozen:
        raise Errors.pool_frozen()
    if pool.currentBalance < sr.amount:
        raise Errors.conflict("Pool balance insufficient at execution time")

    requester = (await session.execute(select(User).where(User.id == sr.requesterId))).scalar_one()

    pool_before = pool.currentBalance
    pool_after = pool_before - sr.amount
    user_before = requester.mainBalance
    user_after = user_before + sr.amount

    pool.currentBalance = pool_after
    requester.mainBalance = user_after

    session.add(Transaction(
        poolId=pool.id, userId=requester.id, type=TransactionType.SPEND,
        direction=TransactionDirection.OUT, amount=sr.amount,
        balanceBefore=pool_before, balanceAfter=pool_after,
        description=f"Spend: {sr.title}", spendRequestId=sr.id,
    ))
    session.add(Transaction(
        userId=requester.id, type=TransactionType.SPEND,
        direction=TransactionDirection.IN, amount=sr.amount,
        balanceBefore=user_before, balanceAfter=user_after,
        description=f"Pool payout: {sr.title}", spendRequestId=sr.id,
        metadata_={"poolId": pool.id},
    ))

    sr.status = SpendStatus.EXECUTED
    await session.commit()
    await session.refresh(sr)
    return {"spendRequest": sr, "poolBalance": pool_after, "userBalance": user_after}


async def cancel_spend_request(session: AsyncSession, *, pool_id: str,
                               spend_request_id: str, actor_id: str) -> SpendRequest:
    sr = (await session.execute(
        select(SpendRequest).where(SpendRequest.id == spend_request_id)
    )).scalar_one_or_none()
    if not sr or sr.poolId != pool_id:
        raise Errors.not_found("Spend request")
    if sr.requesterId != actor_id:
        raise Errors.forbidden("Only requester can cancel")
    if sr.status != SpendStatus.PENDING:
        raise Errors.conflict(f"Cannot cancel spend in {sr.status.value} state")
    sr.status = SpendStatus.CANCELLED
    sr.resolvedAt = datetime.now(timezone.utc)
    await session.commit()
    await session.refresh(sr)
    return sr


async def expire_stale_requests(session: AsyncSession) -> int:
    now = datetime.now(timezone.utc)
    stale = (await session.execute(
        select(SpendRequest).where(
            SpendRequest.status == SpendStatus.PENDING, SpendRequest.expiresAt < now
        )
    )).scalars().all()
    if not stale:
        return 0
    for sr in stale:
        sr.status = SpendStatus.EXPIRED
        sr.resolvedAt = now
        session.add(Notification(
            userId=sr.requesterId, type=NotificationType.SPEND_REQUEST_EXPIRED,
            title="Spend request expired", body=sr.title,
            metadata_={"poolId": sr.poolId, "spendRequestId": sr.id},
        ))
    await session.commit()
    return len(stale)
